Allow in-place operations on views

colesbury opened the pull request:
```
Allow in-place operations on views

Adds VariableViewImpl, a subclass of VariableImpl which has a pointer to
the base Variable on which it is a view. In-place operations on views
change the grad_fn of the base.

Fixes #3313
```

colesbury committed:
Allow in-place operations on views

Adds VariableViewImpl, a subclass of VariableImpl which has a pointer to
the base Variable on which it is a view. In-place operations on views
change the grad_fn of the base.

Note that in-place operations on views from Python-implemented autograd functions
are not currently supported, but in-place operations on non-views still
work.

Fixes #3313

soumith commented:
retest this please

soumith commented:
@pytorchbot retest this please

apaszke commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -381,90 +378,104 @@ static void _transplant_var(VariableImpl& var, const std::shared_ptr<Function>&
+  auto get_shared_base = [&shared_pairs, &t2var](PyObject* tensor) -> Variable {
+    auto input_it = t2var.find(tensor);
+    if (input_it != t2var.end()) {
+      // If the output is an input, it's a view on itself
```
The fact that it's in `t2var` inside this function doesn't mean it has to be an input - it can be one of the previous outputs. Not sure if this changes the logic

apaszke commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -381,90 +378,104 @@ static void _transplant_var(VariableImpl& var, const std::shared_ptr<Function>&
+  };
+
+  // Creates a new Variable from a PyObject* of type THPTensor
+  auto new_variable = [&get_shared_base, is_volatile, cdata](PyObject* output) -> Variable {
```
`cdata` could be captured by reference to avoid a refcount bump

apaszke commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -381,90 +378,104 @@ static void _transplant_var(VariableImpl& var, const std::shared_ptr<Function>&
     }
+
+    // Wrap new tensors in Variables
+    if (!var.defined()) {
```
this could be an `else` of the `if` above

apaszke commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -381,90 +378,104 @@ static void _transplant_var(VariableImpl& var, const std::shared_ptr<Function>&
+    }
+    var.is_volatile() = is_volatile;
+    if (!is_volatile) {
+      var.grad_fn() = cdata;
```
won't this assign a `grad_fn` to things that don't require grad (sth we no longer do for C++ functions)?

apaszke commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -381,90 +378,104 @@ static void _transplant_var(VariableImpl& var, const std::shared_ptr<Function>&
+        // because if someone registers hooks on it, they will fire with grads
+        // from all usages, not only from usages of this output. This is why
+        // we return a view on the input Variable.
+        var = new_variable(output);
```
It seems to me that if we get an input Variable that wasn't a leaf **and** wasn't modified, then it won't enter any of those cases. Can you add a comment to show that we thought about this case as well (there was one for it in the old code)?

apaszke commented on torch/csrc/autograd/variable.h:
```diff
@@ -92,16 +97,20 @@ struct VariableImpl : public at::TensorImpl {
 
 public:
   std::shared_ptr<Function> get_grad_accumulator();
+  virtual std::shared_ptr<Function>& get_grad_fn() {
```
it's a bit annoying that now `grad_fn` lookups are more expensive than for all other attrs

apaszke commented on torch/csrc/autograd/variable.h:
```diff
@@ -114,6 +123,15 @@ struct VariableImpl : public at::TensorImpl {
+
+  Variable base;
+  int expected_version;
+};
```
Can you add a comment briefly describing the view mechanism and mentioning which `VariableImpl` fields might be stale under what conditions?

apaszke commented on torch/csrc/autograd/variable.cpp:
```diff
@@ -87,6 +89,37 @@ std::shared_ptr<Function> VariableImpl::get_grad_accumulator() {
+  }
+  is_view = true;
+  version_counter = base.version_counter();
+  expected_version = version_counter.current_version();
```
I think sth like `attr_version` would describe it better (it's a version that was used to compute the attrs)

apaszke commented on torch/csrc/autograd/variable.cpp:
```diff
@@ -87,6 +89,37 @@ std::shared_ptr<Function> VariableImpl::get_grad_accumulator() {
+}
+
+std::shared_ptr<Function>& VariableViewImpl::get_grad_fn() {
+  std::lock_guard<std::mutex> lock(grad_accumulator_lock);
```
can we rename `grad_accumulator_lock` to `var_lock` or sth like this? It seems unrelated to `grad_accumulator` here

apaszke commented on torch/csrc/autograd/variable.cpp:
```diff
@@ -87,6 +89,37 @@ std::shared_ptr<Function> VariableImpl::get_grad_accumulator() {
+    fn->self_geometry = TensorGeometry(base);
+    fn->size = sizes();
+    fn->stride = strides();
+    fn->storage_offset = Variable(this, true).storage_offset();
```
what's happening here? can't you take `data.storage_offset()`?

apaszke commented on torch/csrc/autograd/variable.cpp:
```diff
@@ -87,6 +89,37 @@ std::shared_ptr<Function> VariableImpl::get_grad_accumulator() {
+    fn->set_flags(Function::flags({ base }));
+    fn->num_inputs = 1;
+    _grad_fn = std::move(fn);
+    expected_version = version_counter.current_version();
```
It seems that there might be a race in here (you read some version, but it changes while you're updating the attrs), although I think it would only happen if someone modifies the base Variable, while you're reading it, which is an error itself. Is that right?

Also, it might be better to get `current_version` once, since it's an atomic load

apaszke reviewed (commented):
Looks good for the most part, but I think there are a few things that should be changed before this is merged (unless I misunderstood some parts of the code)

apaszke commented:
BTW do we still need `live_refs` in `VariableVersion`?

colesbury committed:
Changes from review and rework _wrap_outputs a bit

colesbury commented:
I've changed the logic in `_wrap_outputs` some more. It's worth looking over carefully.

colesbury commented:
I'm still using `live_refs` to prevent in-place operations on views from Python autograd functions. We might be able to change it so that we look at `base` instead.

colesbury committed:
Allow modifications of views from Python autograd functions

 - Python functions can modify views if they return only the single
   modified variable
 - Move shared code to variable.cpp
 - Remove live_refs

apaszke commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -524,30 +536,23 @@ static void _join_version_counters(THPFunction *self, t2var_type &t2var)
+        "and output tensors, but argument %d doesn't satify this "
+        "condition", i);
+
+    map.emplace(t2, it->second);
```
Ohhh I didn't notice it's output -> input map, not the other way around. My bad

apaszke commented on torch/csrc/autograd/variable.cpp:
```diff
@@ -121,6 +122,20 @@ std::shared_ptr<Function>& VariableViewImpl::get_grad_fn() {
 
+void VariableViewImpl::set_grad_fn(std::shared_ptr<Function> grad_fn) {
+  if (this->_grad_fn && grad_fn) {
+    // If we have a previous grad_fn then this is an in-place modification
```
Views don't have grad functions if they don't require grad, right? So you can have in-place modifications even though you didn't have grad_fn, and the fact that you don't have it doesn't imply that your base still doesn't require grad (it might have been modified in the meantime and you didn't check)

colesbury replied:
If a view doesn't have a grad_fn, then it's a leaf and we don't currently support in-place operations on leaves.

The logic here worries me too, though. Perhaps we should explicitly have VariableViewImpl::rebase(grad_fn),  which is only called for in-place modifications, instead of trying to determine if it's a modification or new grad_fn internally.

apaszke replied:
We do support in-place on leaves if the don't require grad. So how is it finally, is it true that views always have grad_fns or not (my guess is no)?

apaszke replied:
Yeah, this situations that turn on requires_grad in base seem really dangerous. Any operations on the views will see that they don't require grad, so they won't emit gradient for them, but they should reach the base!

apaszke reviewed (commented):
In general looks good, but I have a few concerns about correctness. It's really changing the critical parts of autograd and I'm starting to get nervous that it's this large 😕

colesbury commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -382,18 +364,19 @@ static void _wrap_outputs(THPFunction *self, t2var_type &t2var,
     PyObject *raw_output, PyObject *outputs, bool is_volatile)
 {
-  auto cdata = is_volatile ? nullptr : THPFunction_asFunction(self);
+  bool is_executable = self->cdata.is_executable && !is_volatile;
```
self->cdata.is_executable should be sufficient

colesbury commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -404,80 +387,78 @@ static void _wrap_outputs(THPFunction *self, t2var_type &t2var,
-    if (!output_var) throw python_error();
+      // We already have the data tensor wrapped as a PyObject*
+      Py_INCREF(output);
+      Py_XDECREF(output_var->data);
```
Done

colesbury commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -404,80 +387,78 @@ static void _wrap_outputs(THPFunction *self, t2var_type &t2var,
-    return var;
+    // An input has been returned, but it wasn't modified. Return it as a view
+    // so that we can attach a new grad_fn to the Variable.
+    return make_variable_view(std::move(prev), std::move(data));
```
1) Yeah, that should be fixed. test_function_returns_input tests that case

2) Yeah. I think there's a few other simple special case. I'd prefer doing it in a subsequent commit.

colesbury commented on torch/csrc/autograd/variable.cpp:
```diff
@@ -121,6 +122,20 @@ std::shared_ptr<Function>& VariableViewImpl::get_grad_fn() {
+    // If we have a previous grad_fn then this is an in-place modification
+    if (output_nr != 0 || grad_fn->num_inputs != 1) {
+      throw std::runtime_error("Functions which modify views in-place must return a single Variable");
+    }
```
I forgot to enforce `output_nr` = 0 for views. I'll add those assertions.

colesbury committed:
Changes from latest review

colesbury committed:
Split set_history into set_history and rebase_history and fix in-place
NN functions

colesbury commented:
Rebased to get TORCH_ASSERT and added two more commits

colesbury commented:
Here's the diff for the generated VariableType.cpp:

https://gist.github.com/colesbury/bc6835aaef3f755dd063cb484e788701

colesbury committed:
Fix crash in CopySlices when should_compute_output(0) is false

apaszke commented on torch/csrc/autograd/functions/tensor.cpp:
```diff
@@ -159,12 +159,18 @@ auto CopySlices::apply(const variable_list& inputs) -> variable_list {
-    grad_inputs[i] = std::move(res[i]);
+  for (size_t i = 0; i < res.size(); i++) {
+    if (should_compute_output(i)) {
+      TORCH_ASSERT(res[i].defined());
```
I think this assertion is too strict, mostly because `should_compute_output` is really an approximation, that might sometimes have false positives. Right now it's really a check of `is_executable` of `i`th next function. However, it is possible that a function is executable, but its `i`th output was non-differentiable (e.g. it was a max index), and it was used as a target for the loss. This loss will check that its input didn't require grad at runtime, and its backward won't return grad for it.

apaszke commented on test/test_autograd.py:
```diff
@@ -1542,6 +1542,14 @@ def func(root, b):
         go = Variable(torch.randn(a.size()), requires_grad=True)
         gradgradcheck(func, (a, b), (go,))
 
+    def test_inplace_view5(self):
```
Can you please add docstrings to these tests? `test_inplace_view83` won't really tell you what's the essence of the problem and no one will remember that soon. Also, one line docstrings show up in test errors and verbose outputs, which is really nice

apaszke commented on tools/autograd/gen_variable_type.py:
```diff
@@ -771,19 +768,26 @@ def emit_body(declaration):
+            if is_view:
+                base_call = 'auto ret = as_view(static_cast<const Variable&>(self), {})'.format(base_call)
+            else:
+                base_call = 'auto ret = as_variable({})'.format(base_call)
```
Is there any significant change in the logic here, or have you just rearranged things? Is it the addition of `maybe_wrap` (and why is it necessary)?

apaszke commented on torch/csrc/autograd/variable.h:
```diff
@@ -56,7 +56,10 @@ struct Variable : public at::Tensor {
   inline       std::shared_ptr<Function>& grad_fn();
 
+  // Sets the flags and grad_fn ("history") of a new Variable
   inline void set_history(VarFlags flags, int output_nr, std::shared_ptr<Function> grad_fn);
```
I really thing this would have been much cleaner and safer if we just used the constructor for `set_history` since it's not new Variables only, and `rebase_history` for all other cases

apaszke commented on torch/csrc/autograd/variable.h:
```diff
@@ -218,14 +220,17 @@ inline std::shared_ptr<Function>& Variable::grad_fn() {
   return get()->get_grad_fn();
 };
 inline void Variable::set_history(VarFlags flags, int output_nr, std::shared_ptr<Function> grad_fn) {
+  assert(!get()->_grad_fn && "set_history can only be called on new Variables");
```
`TORCH_ASSERT`. This would never get triggered because non-debug builds of Python build extensions with `-DNDEBUG`

apaszke commented on tools/autograd/templates/VariableType.h:
```diff
@@ -53,6 +53,7 @@ struct VariableType : public at::Type {
   std::tuple<Variable, Variable> as_variable(std::tuple<Tensor, Tensor> tensor) const;
   std::tuple<Variable, Variable, Variable> as_variable(std::tuple<Tensor, Tensor, Tensor> tensor) const;
   std::vector<Variable> as_variable(TensorList tensor) const;
+  Variable maybe_wrap(Tensor data, const Variable & self, bool inplace) const;
```
I can't see a reason why would this need to be a `VariableType` method (it doesn't even use `this`) and not a static helper in the file. I think it would be clearer that way

colesbury replied:
It calls `as_variable` which is on `VariableType`. `as_variable` has to be on VariableType because the overload for `Scalar`uses `this->baseType`. C++ doesn't allow a mix of overloads (`as_variable`) that are methods and functions.

colesbury replied:
This can be simplified once ATen only returns "Tensor" instead of "Scalar"

apaszke commented on torch/csrc/autograd/variable.cpp:
```diff
@@ -122,18 +124,22 @@ std::shared_ptr<Function>& VariableViewImpl::get_grad_fn() {
+  TORCH_ASSERTM(requires_grad, "Can't set grad_fn on view with requires_grad=False");
+  TORCH_ASSERT(output_nr == 0);
+  if (grad_fn->num_inputs != 1) {
+    throw std::runtime_error("Functions which modify views in-place must return a single Variable");
```
Is that right? I thought the only limitation was that views have to be the first output. Or is it because `CopySlices` would have to handle that?

colesbury replied:
Yeah, `CopySlices` would have to handle that, which is difficult and doesn't seem worth it for now at least.

apaszke commented on torch/csrc/autograd/variable.h:
```diff
@@ -42,9 +51,16 @@ struct Variable : public at::Tensor {
+  inline bool is_leaf() const;
+
   inline const std::shared_ptr<Function>& grad_fn() const;
   inline       std::shared_ptr<Function>& grad_fn();
```
Can we make it always return a const shared pointer? I think all modifications to `grad_fn` should be through `set_history` and `rebase_history`, right?

apaszke commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -382,87 +356,101 @@ static void _transplant_var(VariableImpl& var, const std::shared_ptr<Function>&
+    PyObject *raw_output, PyObject *outputs, bool is_volatile)
 {
-  auto cdata = is_volatile ? nullptr : THPFunction_asFunction(self);
+  bool is_executable = self->cdata.is_executable && !is_volatile;
```
Can you remove `!is_volatile`? The first condition should be enough (as discussed in the previous review)

apaszke reviewed (commented):
Looks good, but we should discuss the situation when `requires_grad` of `base` changes

colesbury committed:
Add assertion that the requires_grad in the base and view when the view
is used in a differntiable operation.

colesbury committed:
Remove non-const overload of grad_fn()

Record base in SavedVariable. We were not saving the "base" in
SavedVariable. This was probably a bug, but no detectable because we
don't do in-place operations on saved variables.

colesbury committed:
Add private accessor for _base and use try/catch

colesbury committed:
Change detach() to not return a view

The detach() function still shares version counters with the original
Variable, but doesn't return a view.

apaszke commented on torch/csrc/autograd/variable.cpp:
```diff
@@ -108,6 +108,14 @@ VariableViewImpl::VariableViewImpl(Variable base_, at::Tensor data_)
+    throw std::runtime_error(
+        "requires_grad is False and base.requires_grad is True. Cannot use "
+        "this view in a differentiable operation. Re-create the view from the "
+        "base Variable.");
```
It might be a bit more helpful to say "re-create the view form its base Variable after the last in-place modification".

apaszke commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -391,6 +431,7 @@ static struct PyGetSetDef THPVariable_properties[] = {
   {"data", (getter)THPVariable_get_data, (setter)THPVariable_set_data, NULL, NULL},
   {"_grad", (getter)THPVariable_get_grad, (setter)THPVariable_set_grad, NULL, NULL}, // only for legacy reasons
   {"grad", (getter)THPVariable_get_grad, (setter)THPVariable_set_grad, NULL, NULL},
+  {"_base", (getter)THPVariable_get_base, NULL, NULL, NULL}, // only for legacy reasons
```
what legacy reasons?

colesbury replied:
oops - copy pasta

apaszke reviewed (commented):
Last commits look good to me, but please add more comments to the tests to clearly state what you're checking. I need one more day to look go over `VariableViewImpl` again and ensure that there are no more weird edge cases.

colesbury committed:
Improve error message and remove incorrect comment

apaszke commented on torch/csrc/autograd/variable.cpp:
```diff
@@ -87,6 +91,65 @@ std::shared_ptr<Function> VariableImpl::get_grad_accumulator() {
+  }
+  is_view = true;
+  version_counter = base.version_counter();
+  attr_version = version_counter.current_version();
```
I think it would be safer to set `attr_version` to an invalid value like -1. You didn't set up a `grad_fn` in here, but it might be necessary, so we better enter the slow path of `get_grad_fn()` next time

colesbury replied:
I've changed the constructor so it takes a grad_fn

apaszke commented on torch/csrc/autograd/variable.cpp:
```diff
@@ -87,6 +91,65 @@ std::shared_ptr<Function> VariableImpl::get_grad_accumulator() {
+}
+
+std::shared_ptr<Function>& VariableViewImpl::get_grad_fn() {
+  std::lock_guard<std::mutex> lock(mutex);
```
It seems to me that it would be enough to acquire the mutex in the slow path.

colesbury replied:
The read on `_grad_fn` should be guarded by the mutex, since another call to `get_grad_fn` could modify it.

apaszke commented on torch/csrc/autograd/variable.cpp:
```diff
@@ -87,6 +91,65 @@ std::shared_ptr<Function> VariableImpl::get_grad_accumulator() {
+  }
+
+  TORCH_ASSERTM(requires_grad, "Can't set grad_fn on view with requires_grad=False");
+  TORCH_ASSERT(output_nr == 0);
```
Can you move this assert to the beginning, so it is always checked?

apaszke replied:
Also, shouldn't we assert that `base.requires_grad()` matches the value of this function?

apaszke commented on torch/csrc/autograd/variable.cpp:
```diff
@@ -87,6 +91,65 @@ std::shared_ptr<Function> VariableImpl::get_grad_accumulator() {
+  auto copySlices = std::make_shared<CopySlices>(base, TensorGeometry(data), std::move(grad_fn));
+  base.output_nr() = 0;
+  base.get()->_grad_fn = std::move(copySlices);
+  base.requires_grad() = true;
```
This line depends on the check that in-place functions return only a single value. Can you please add a warning or sth?

colesbury replied:
There's an assertion for that immediately above this.

apaszke replied:
I know, but it's not immediately clear that this line depends on the assertion. Adding a tiny comment saying that it is like this won't hurt, and we won't miss it in the future for sure

colesbury replied:
What line are you referring to? The `base.output_nr() = 0` doesn't depend on that assertion.

apaszke commented on torch/csrc/autograd/variable.h:
```diff
@@ -155,12 +209,26 @@ inline Variable & Variable::grad() {
+  return get()->get_grad_fn();
 };
+inline void Variable::set_history(VarFlags flags, int output_nr, std::shared_ptr<Function> grad_fn) {
+  assert(!get()->_grad_fn && "set_history can only be called on new Variables");
```
Use `TORCH_ASSERT`. Plus, this doesn't really imply that it's a new Variable (it might be an old Variable that only now starts requiring grad).

apaszke replied:
Also, can you add an assertion that `grad_fn` is null if `flags.requires_grad` is false? Same below

colesbury replied:
We still create grad_fn for Python-implemented functions that don't require grad. (I'm going to change that in a subsequent PR).

apaszke commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -382,87 +356,102 @@ static void _transplant_var(VariableImpl& var, const std::shared_ptr<Function>&
+          torch::createTensor(output),
+          get_shared_base(output),
+          is_modified);
+      if (is_modified) {
```
Correctness of these lines depends on the code of `wrap_output` which is a bit far right now. Can you add a comment saying that it's safe to just call `set_history`, because the only case where we return the original variable is when it was an input and modified, so it will always enter the first case?

colesbury replied:
I've moved most of `set_history` into `make_variable` and moved the rebase call up to wrap_output.

apaszke reviewed (commented):
Alright, I think I'm ok with merging this once you address these comments. Please please please please add the test descriptions as well, I have no idea what are they checking.

colesbury committed:
Fixes from review:

 - Rename and add descriptions to autograd tests
 - Move most of set_history to make_variable constructors
 - Additional assertions

colesbury commented:
I've removed the calls to `set_history` in-favor of passing those arguments to Variable factory functions. The generated VariableType.cpp still effectively calls "set_history", but it's inlined into the calling code. In a subsequent PR, I'll clean up VariableType.cpp. I don't want to do it here because that change is probably substantial and this isn't a correctness issue.

apaszke commented on torch/csrc/autograd/variable.cpp:
```diff
@@ -91,8 +86,9 @@ std::shared_ptr<Function> VariableImpl::get_grad_accumulator() {
-  : VariableImpl(std::move(data_))
+VariableViewImpl::VariableViewImpl(Variable base_, at::Tensor data_, VarFlags flags,
+                                   int output_nr, std::shared_ptr<Function> grad_fn)
+  : VariableImpl(std::move(data_), flags, output_nr, std::move(grad_fn))
```
nit: we get `output_nr` now, so add `TORCH_ASSERT(output_nr == 0)`

apaszke commented on torch/csrc/autograd/variable.h:
```diff
@@ -164,20 +164,20 @@ struct VariableViewImpl : public VariableImpl {
 
 inline Variable make_variable(at::Tensor data, bool requires_grad, bool is_volatile=false) {
-  return Variable(new VariableImpl(std::move(data), requires_grad, is_volatile), false);
+  return make_variable(std::move(data), VarFlags(requires_grad, is_volatile));
```
this overload is a bit weird now, as everything else uses `VarFlags`. but I don't mind having it

colesbury replied:
Yeah, we should clean this up soon

apaszke reviewed (approved):
LGTM

apaszke commented:
Actually you haven't added a check for the case when we're rebasing a view that requires grad (we should check that base still requires grad as well)

colesbury committed:
Add test that modifying view in-place makes the base require grad

colesbury committed:
Revert output_nr assertion in VariableViewImpl constructor

 - Views may have output_nr != 0, but then we don't allow in-place
   modifications to them
